knotx-core/src/main/java/io/knotx/assembler/FragmentAssemblerFallbackHandler.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2016 Cognifide Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.knotx.assembler;
+
+import com.google.common.collect.Maps;
+import io.knotx.dataobjects.Fragment;
+import io.knotx.dataobjects.KnotContext;
+import io.knotx.fallback.DefaultFallbackStrategy;
+import io.knotx.fallback.FallbackStrategy;
+import io.knotx.fragments.FragmentConstants;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import java.util.Map;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import org.apache.commons.lang3.StringUtils;
+
+public class FragmentAssemblerFallbackHandler {
+
+  private static final Logger LOGGER = LoggerFactory
+      .getLogger(FragmentAssemblerFallbackHandler.class);
+
+  private final Map<String, FallbackStrategy> fallbackStrategies = Maps.newHashMap();
+
+  private final Map<String, Fragment> fallbackFragmentCache = Maps.newHashMap();
+
+  private final FragmentAssemblerOptions options;
+
+  public FragmentAssemblerFallbackHandler(FragmentAssemblerOptions options) {
+    this.options = options;
+    ServiceLoader<FallbackStrategy> fallbackStrategyServiceLoader = ServiceLoader.load(FallbackStrategy.class);
+    for (FallbackStrategy strategy : fallbackStrategyServiceLoader) {
+      fallbackStrategies.put(strategy.getId(), strategy);
+    }
+  }
+
+  public String applyFallback(Fragment failed, KnotContext knotContext) {
+    Fragment fallback = getFallback(failed, knotContext);
+    FallbackStrategy strategy = getFallbackStrategy(fallback);
+    return strategy.applyFallback(failed, fallback, knotContext);
+  }
+
+  private FallbackStrategy getFallbackStrategy(Fragment fallbackFragment) {
+    String strategyId = Optional.ofNullable(fallbackFragment.getAttribute(FragmentConstants.FALLBACK_STRATEGY))
+        .orElse(DefaultFallbackStrategy.ID);
+    return Optional.ofNullable(fallbackStrategies.get(strategyId)).orElseThrow(() -> {
+      LOGGER.error("Fragment {} specifies fallback strategy but no fallback strategy with given id was found", fallbackFragment);
+      return new IllegalStateException(String.format("no strategy with id %s found", strategyId));
```

Comment:
This is just cosmetics, but I'm not sure this is proper exception used here.
I see `IllegalStateException` as the wrong time to do some logic. Maybe `IllegalArgumentException` would be better fit here, what do you think?

Reply:
I agree, changed to `IllegalArgumentException`